### DISCUSSION TITLE
The "End" of Aliases

### DISCUSSION THREAD

**Discussion by APickledWalrus** [SkriptLang/Skript#4164] [Suggestions]
As Skript prepares to drop support for 1.12 and below after 2.6, the lack of a need for Skript's current alias and item/block handling system is becoming more and more apparent.

The idea of removing aliases, specifically classes in the aliases package such as ItemType and ItemData is nothing new. While these classes have been beneficial for managing items, blocks, and comparisons (even though it has not always worked so well), the modern API features are more than sufficient. These classes and comparisons have become little more than chore for maintainers to manage. The API already includes comparison methods for many of these things (blocks, items, etc.)

Features like block data, which continue to be integrated into Skript, are ideal replacements for this system. With block data, Skript does not need to worry about handling block comparisons or creating aliases for the different permutations/block states of a block. Plus, it makes it much easier on the end user. If we use Minecraft's own systems (e.g. `cobblestone_wall[north=true,south=true]`, many users may already be familiar, or can even learn how to work these systems with Minecraft and Skript simultaneously. 

With block data integrated, Skript can separate blocks and items, making things far easier to manage. Realistically, only the creation of basic item IDs would need to be handled by Skript. To make this even easier, we already have basic item IDs: The IDs provided by Minecraft itself for items. By expanding on PRs like https://github.com/SkriptLang/Skript/pull/4034, we no longer have to worry about alias creation whatsoever.

The only thing that would really remain is data tags for items. We could handle that the way it's done in aliases or in game now:
Example:
```
add minecraft:potion{"Potion": "minecraft:slow_falling"} to the player's inventory
```

This solution does have some cons though. See the table below for general pros/cons:
| Pros | Cons |
| - | - |
| Less work is required by maintainers | Will require a lot of work to detach/debug everything |
| Consistent with Minecraft's existing systems | Will require work for users to convert their scripts |
| Easier for users to know how to use items/blocks (I often see people confused about alias names/variations)

This "proposal" was just a place for me to put down my ideas.  know this is something the community/devs have discussed before, but this would be a good place to keep thoughts. I would like to hear what everyone else thinks about this matter. I know we are already planning something similar to this with the replacement of the current string formatting system through the usage of MiniMessage.

Thanks for reading :D

**Comment by AyhamAl-Ali** [comment]
I always liked the aliases system but when I started contributing and saw how painful it is to keep up with the new items every update and fix aliases in some places that doesn't work, I think this is a good idea for the future of Skript but indeed it will be a bit pain at the beginning to keep testing the system until proven safe and for the users to convert.

However, I like the idea.

**Comment by TheBentoBox** [comment]
I could see a route forward where the aliases system is removed or at least drastically simplified. My main two cents is that I've seen plenty of good suggestions denied or argued against because Skript is supposed to be plain-English. I don't see how `minecraft:potion{"Potion": "minecraft:slow_falling"}` is plain-English.

The real main point of the aliases is not simply to create a way to refer to items; it's to allow players to refer to items, individually or categorically, more naturally. Stuff like `if player's tool is any pickaxe`, or (in reference to the aforementioned potion example), `give player a potion of slow falling`.

At the end of the day it matters little to me, but if you guys actually care about allowing stuff to still be written in plain-English I don't think actually removing aliases is a viable option. Drastically slimming them down is another story though; I can certainly see them being reduced to only being used for item categories (i.e. `any X`), or for complex items like potions which need NBT data to be properly described. Then the simplistic aliases (which make up the very large majority of items) could be generated automatically. I believe a hybrid approach like that is the best one, but up to you guys in the end.

**Reply by APickledWalrus** [reply]
Thank you for your input. I certainly agree with you on the point of more complex aliases like potions - where it would be to the user's benefit to have manually created aliases

**Comment by Moderocky** [comment]
I think there are pros and cons to both approaches, and it may be possible to achieve a balance that gets the benefits of both.

Counter-intuitively, I expect that using the block-data / item data formats (e.g. `stone_slab[type=top` or `iron_sword{Damage:1}`) will actually be **easier** for new users than the current aliases: it's the format used in the built-in minecraft commands and displayed in the block tooltips, so users with any existing minecraft experience will understand the format.

That said, for some people who have never touched commands, command-blocks, resource packs etc. and are not used to their format, the existing aliases system might be simpler (at least for basic aliases, the odd ones perhaps not.)

The current aliases system could be maintained separately and offered as an alternative: users who feel they don't need it can switch it off in the config to make parsing, etc. faster, while users that benefit from it can leave it on. This would satisfy everybody and harm nobody - aliases could be switched on by default so no backwards compatibility issues.

**Reply by Mwexim** [reply]
I concur, since 'normally' you don't really use these special types anyway. The only 'data'-altering modifier when using materials is the data value (difference between oak planks and spruce planks for example), but that is completely erased if you drop support for 1.12 and lower.

The system seems so frustrating to implement and I feel that it can be automated completely by using the proposed system. Also, aliases can become so ridiculously long and complicated that they don't even make sense anymore.

Even if this change is not English-y enough, I can see Skript developing some really smart system to handle aliases:
- It automatically detects group patterns (for example, it sees `OAK_PLANKS`, `SPRUCE_PLANKS`, etc. and will therefore create an alias `any planks`)
- It can automatically handle data classes for materials (for example, it adds `[powered]` to Powerable materials like levers and buttons)

**Comment by ShaneBeee** [comment]
Im going to make my reply short and sweet.
I think we need to at LEAST remove "block state" aliases, ie (btw im making this up as I dont know the real ones):
`set target block to north facing upside down corner oak stairs` <--- NO ONE USES THESE

As stated in your argument, since we now support block data (for quite some time now), THESE aliases are virtually useless, and really, they're the biggest headache of them all. 

I do feel some aliases are good to keep, ie: potions.
Some others are great as well, such as groupings (any sword, any chest plate, any plant)

Another note, in a recent release, we added "temp aliases", which basically loops Bukkit's Materials, and creates aliases for missing items. That said, we could theoretically remove 99% of the aliases, let Skript create these "temp" aliases, and let uses use block data.
This would GREATLY cut down Skript loading time, currently: 
On Paper: 
```
[17:48:02 INFO]: [Skript] Loaded 143127 aliases in 4544ms
```
On Spigot:
```
[18:36:30] [Server thread/INFO]: [Skript] Loaded 143127 aliases in 13604ms
```
#OUCHIE 
143k+ aliases, at an average of 4 - 14 seconds, is just painful on a server. 

On top of all this, each update (with the exception of 1.18), Mojang is adding a ton of new blocks/items, which makes aliases a LOT of work on the devs, as well as an insane growing time for alias parsing during server load. 

By cutting out all block state aliases, and the majority of aliases that match Material names, this would most likely not affect the majority of Skript users. Some will obviously need to update their scripts, while others wont even notice a difference. 

This could be a good step in the right direction to eventually remove Aliases all together.

**Comment by TheBentoBox** [comment]
Just my two cents as the one who wrote the majority of the baseline aliases for the new alias system, I'd be happy to see them gone in favor of a more automated system. I agree nobody really uses those extremely complex aliases, and even if they did the fact that the order matters is a ridiculous headache (if the alias is defined as `<waterlogged> <shape> <material>`, you can't do `<shape> <waterlogged> <material>`). It's really just not intuitive and something more flexible like the block data makes much more sense long-term.

**Comment by lewis252310** [comment]
As a foreign language user, I think this is a good idea, but don't delete it completely.
Of course, I hope it will be a mix mode that can be used by both aliases and NBT.

For some foreign language user, sometimes using aliases is even more difficult to understand than using NBT.
So most of the SK foreign language writers I know prefer to use `NBT(by Skbee)` or `data of block` directly for editing. Like the attributes and enchantments of the item, or the distance between leaves and logs.

After all, almost all foreign language users who will come into contact with Skript already have the most basic knowledge of NBT, it is a bit painful to switch to normal English at once (Well, this is our own problem :) ). It's even like I am the writer who switched from vanilla Datapack to Skript.
Of course, we need to pay attention to the original intention of "simple English", `[any]` and `[all]` are very good ones, much better than vanilla `#tag`, and they helped me a lot in the early days of learning Skript.
I even learned a lot of basic English grammar from Skript.

Maybe we can make simple English and vanilla NBT parallel, and satisfy different users at the same time, and I think, the writers who know which blocks have directional, or the growth stages of crops should also understand block data.
Obviously they I have left the level of simple English. It is indeed feasible to delete some aliases for this part.

Then I noticed that we seem to have forgotten certain groups? (Or just because I am not familiar with English?)
For example, `slime` is not included in `monsters` (I think this is a small oblivion caused by the Spigot-API class), or the hostility of Mobs cannot be distinguished more intuitively.
But from this it can be seen that the builder of the alias is being troubled by other things, and there is no room to deal with other more basic aliases.

The above is an idea from a foreign language user who likes Skript very much, and I hope it will be helpful to the development team.
But please allow me to apologize for my typography and tone first, and I'm really sorry if I offend. I will work harder to learn English.
`(Using by google translation)`
